src/core/qgsgdalutils.cpp

```diff
@@ -120,6 +186,132 @@ void QgsGdalUtils::resampleSingleBandRaster( GDALDatasetH hSrcDS, GDALDatasetH h
   GDALDestroyGenImgProjTransformer( psWarpOptions->pTransformerArg );
 }
 
+gdal::dataset_unique_ptr QgsGdalUtils::resampleImage( const QImage &image, const QgsRectangle &extent, QSize outputSize, GDALResampleAlg resampleAlg )
+{
+  gdal::dataset_unique_ptr srcDS = QgsGdalUtils::imageToMemoryDataset( image, extent, QgsCoordinateReferenceSystem() );
```

Comment:
This is rather inefficient to use the warp algorithm to do just resampling without reprojection (this applies to the above resampleSingleBandRaster() method). You could get even better results with using GDALRasterIOEx() with the GDALRasterIOExtraArg.eResampleAlg